src/distributions/weighted.rs

```diff
@@ -130,6 +132,342 @@ impl<X> Distribution<usize> for WeightedIndex<X> where
+            let s = aliases.pop_small();
+            let b = aliases.pop_big();
+
+            aliases.set_alias(s, b);
+            no_alias_odds[b] = no_alias_odds[b] - weight_sum + no_alias_odds[s];
+
+            if no_alias_odds[b] < weight_sum {
+                aliases.push_small(b);
+            } else {
+                aliases.push_big(b);
+            }
+        }
+
+        // The remaining indices should have no alias odds of about 100%. This is due to
+        // numeric accuracy. Otherwise they would be exactly 100%.
+        while !aliases.smalls_is_empty() {
+            no_alias_odds[aliases.pop_small()] = weight_sum;
+        }
+        while !aliases.bigs_is_empty() {
+            no_alias_odds[aliases.pop_big()] = weight_sum;
+        }
+
+        // Prepare distributions for sampling. Creating them beforehand improves
+        // sampling performance.
+        let uniform_index = super::Uniform::new(0, n);
+        let uniform_within_weight_sum = super::Uniform::new(W::ZERO, weight_sum);
+
+        Ok(Self {
+            aliases: aliases.aliases,
+            no_alias_odds,
+            uniform_index,
+            uniform_within_weight_sum,
+        })
+    }
+}
+
+impl<W: AliasMethodWeight> Distribution<usize> for AliasMethodWeightedIndex<W> {
+    fn sample<R: Rng + ?Sized>(&self, rng: &mut R) -> usize {
+        let candidate = rng.sample(self.uniform_index);
+        if rng.sample(&self.uniform_within_weight_sum) < self.no_alias_odds[candidate] {
+            candidate
+        } else {
+            self.aliases[candidate]
+        }
+    }
+}
+
+impl<W: AliasMethodWeight> fmt::Debug for AliasMethodWeightedIndex<W>
+where
+    W: fmt::Debug,
+    super::Uniform<W>: fmt::Debug,
+{
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_struct("AliasMethodWeightedIndex")
+            .field("aliases", &self.aliases)
+            .field("no_alias_odds", &self.no_alias_odds)
+            .field("uniform_index", &self.uniform_index)
+            .field("uniform_within_weight_sum", &self.uniform_within_weight_sum)
+            .finish()
+    }
+}
+
+impl<W: AliasMethodWeight> Clone for AliasMethodWeightedIndex<W>
+where
+    super::Uniform<W>: Clone,
+{
+    fn clone(&self) -> Self {
+        Self {
+            aliases: self.aliases.clone(),
+            no_alias_odds: self.no_alias_odds.clone(),
+            uniform_index: self.uniform_index.clone(),
+            uniform_within_weight_sum: self.uniform_within_weight_sum.clone(),
+        }
+    }
+}
+
+/// In comparision to naive accumulation, the pairwise sum algorithm reduces
+/// rounding errors when there are many floating point values.
+fn pairwise_sum<T: AliasMethodWeight>(values: &[T]) -> T {
+    if values.len() <= 32 {
+        values.iter().map(|x| *x).sum()
+    } else {
+        let mid = values.len() / 2;
+        let (a, b) = values.split_at(mid);
+        pairwise_sum(a) + pairwise_sum(b)
+    }
+}
```

Comment:
Instead of the pairwise sum, [an algorithm proven to be correct](http://www.cs.cmu.edu/~quake-papers/robust-arithmetic.ps) could be used. This is [internally used in libtest](https://github.com/rust-lang/rust/blob/31eb0e2d3c0bfd2e0f5b662d82971c1a203708be/src/libtest/stats.rs#L154). For integers, the naive summation is fine. It might be more efficient to move the implementation of the numerically correct summation to `AliasMethodWeight`, so it can be specialized for integers.

Reply:
The sum does not need to be exact because an approximated sum only reduces the accuracy of the sampling and doesn't make it completely wrong. I just wanted to avoid the effect that the sum doesn't change at all when there are many weights, which for `f32` weights happens pretty early: https://play.rust-lang.org/?version=nightly&mode=release&edition=2018&gist=9b2e8d44c8e4dc780368ecf4791764b1. Pairwise summation, which does have an upper bound on the error too (https://en.wikipedia.org/wiki/Pairwise_summation), avoids that and at the same time doesn't need any more addition operations than naive summation.

In 8641a9bb1624e9a984a69ae0c24fcf16ee846bc8 I changed the code, so that pairwise summation is only used for floating point weights.

Reply:
Fair enough, this seems like a reasonable tradeoff for performance.